python/tvm/topi/arm_cpu/conv2d_alter_op.py

```diff
@@ -121,7 +124,33 @@ def _alter_conv2d_layout(attrs, inputs, tinfos, out_type):
 
     idxd = tvm.tir.indexdiv
 
-    # We don't perform layout alteration for NHWC layout with real data types
+    if topi_tmpl == "depthwise_conv2d_nhwc_dsp.arm_cpu":
+        assert data_layout == "NHWC" and kernel_layout == "HWOI"
+        channels = get_const_tuple(data.shape)[3]
+        KH, KW, _, _ = get_const_tuple(kernel.shape)
+        simd_width = get_dtype_simd_width(data.dtype)
+
+        HWOI_kernel_np = inputs[1].data.numpy()
```

Comment:
You'll need to check that the kernel is a constant and fallback to a different implementation if it is not.

Reply:
I'm not sure how easy it is to check if the kernel is a constant from `python/tvm/relay/op/strategy/arm_cpu.py`, but you're right that it is a thing we should check. I've added an assertion, though it is a bit of a stopgap solution.

Reply:
Could you add a message to the assert and a comment about what needs to be done to not make it a stopgap.

Reply:
Unfortunately, a clean fix is pretty hard. Inside `alter_conv2d_layout`, the type signature allows us to easily check if the kernel is a constant:
```

Reply:
Unfortunately, a clean solution is hard, as the strategy function does not have access to the needed information. When `conv2d_alter_op` is called, `inputs[1]` (the kernel) has the form:
```python
meta[relay.Constant][0] /* ty=Tensor[(3, 3, 3, 8), int16] */
```
However when the Relay strategy functions are called, `inputs[1]` (the kernel) looks like:
```python
Tensor(shape=[3, 3, 8, 1], op.name=placeholder)
```
Nowhere inside `relay/op/strategy` do any of the strategy functions check whether the relevant tensors are constant, so there's not much we can do. I've added comments explaining this, but please let me know if you have ideas for how this could be done.